src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/BookBinder.java

```diff
@@ -0,0 +1,138 @@
+        addItemSetting(bypassVanillaMaxLevel, hasCustomMaxLevel, customMaxLevel); 
+    }
+
+    @Override
+    protected MachineRecipe findNextRecipe(BlockMenu menu) {
+        for (int slot : getInputSlots()) {
+            ItemStack target = menu.getItemInSlot(slot == getInputSlots()[0] ? getInputSlots()[1] : getInputSlots()[0]);
+
+            ItemStack item = menu.getItemInSlot(slot);
+            if  (isCompatible(item) && isCompatible(target)) {
+                
+                EnchantmentStorageMeta itemMeta = (EnchantmentStorageMeta) item.getItemMeta();
+                EnchantmentStorageMeta targetMeta = (EnchantmentStorageMeta) target.getItemMeta();
+
+                Map<Enchantment, Integer> storedItemEnchantments = itemMeta.getStoredEnchants();
+                Map<Enchantment, Integer> storedTargetEnchantments = targetMeta.getStoredEnchants();
+                Map<Enchantment, Integer> enchantments = combineEnchantments(storedItemEnchantments, storedTargetEnchantments);
+                
+                if (enchantments.size() > 0) {
+                    ItemStack book = new ItemStack(Material.ENCHANTED_BOOK);
+                    book.setAmount(1);
+
+                    EnchantmentStorageMeta enchantMeta = (EnchantmentStorageMeta) book.getItemMeta();
+                    for (Map.Entry<Enchantment, Integer> entry : enchantments.entrySet()) {
+                        if (bypassVanillaMaxLevel.getValue()) {
+                            enchantMeta.addStoredEnchant(entry.getKey(), entry.getValue(), true);
+                        } else {
+                            enchantMeta.addStoredEnchant(entry.getKey(), entry.getValue(), false);
+                        }
+                    }
+
+                    book.setItemMeta(enchantMeta);
+                    
+                    MachineRecipe recipe = new MachineRecipe(25 * enchantments.size() / this.getSpeed(), new ItemStack[] {target, item}, new ItemStack[] {book});
+
+                    if (!InvUtils.fitAll(menu.toInventory(), recipe.getOutput(), getOutputSlots())) {
+                        return null;
+                    }
+
+                    for (int inputSlot : getInputSlots()) {
+                        menu.consumeItem(inputSlot);
+                    }
+
+                    return recipe;
+                }
+
+                return null;
+            }
+
+        }
+
+        return null;
+    }
+
+    private boolean isCompatible(@Nullable ItemStack item) {
+        if (item != null && item.getType() == Material.ENCHANTED_BOOK) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public ItemStack getProgressBar() {
+        return new ItemStack(Material.IRON_CHESTPLATE);
+    }
+
+    @Override
+    public String getMachineIdentifier() {
+        return "BOOK_BINDER";
+    }
+
+    private Map<Enchantment, Integer> combineEnchantments(Map<Enchantment, Integer> ech1, Map<Enchantment, Integer> ech2) {
```

Comment:
```suggestion
    @Nonnull
    @ParametersAreNonnullByDefault
    private Map<Enchantment, Integer> combineEnchantments(Map<Enchantment, Integer> ech1, Map<Enchantment, Integer> ech2) {
```

(Nonnull would need to be imported too)